README.md
# homebridge-denon-heos

[![NPM version](https://badge.fury.io/js/homebridge-denon-heos.svg)](https://npmjs.org/package/homebridge-denon-heos)
![License](https://img.shields.io/badge/license-MIT-lightgrey.svg)
[![Downloads](https://img.shields.io/npm/dm/homebridge-denon-heos.svg)](https://npmjs.org/package/homebridge-denon-heos)

## A Plugin for homebridge to control a Denon AVR

This is a [Homebridge](https://github.com/nfarina/homebridge) plugin to control all types of [Denon AVRs](https://www.denon-hifi.nl/nl/product/homecinema/avreceiver) and possibly Marantz AVRs (some are tested with positive results). The plugin uses the http commands as default. As newer Denon receiver don't support http commands anymore, Telnet control is also possible. DUe to the fact that I found these less stable than the http commands, I still prefer the http commands. This git has been forked from [nneubauer/homebridge-denon](https://github.com/nneubauer/homebridge-denon) which was the basis for version 1 of this plugin.

Note: I'm not a javascript coder at all, so be easy on code feedback ;), though please if you have feedback or ideas, let me know.


## Install

The plugin registers as [homebridge-denon-heos](https://www.npmjs.com/package/homebridge-denon-heos). I found that the original [homebridge-denon](https://www.npmjs.com/package/homebridge-denon) and all other plugins I tested were not working with newer Denon AVR generations (mine is X1400) or used the unstable telnet commands. I'm aksing you to post an issue if you have problems or an idea about this plugin. You can install the package with the following command:

```
(sudo) npm install -g homebridge-denon-heos
```

For a specific version number use (change the number after the `@` to the desired version):
```
(sudo) npm install -g homebridge-denon-heos@2.0.0
```
The changelog of the version can be found in the the git.

## Idea

Since there is no HomeKit AVR accessory type, [nneubauers](https://github.com/nneubauer) idea was to create bunch of switches. Each switch signifies an input source which I forked to a version I use now for over a year without any problems. Now with the TV accessories in Homekit, I wanted to make it possible to control my receiver with a TV accessory block and the control widget.


### TV Accessories
Since version 2.0, this plugin is extended with support for TV accessories. This means that you can add your receiver as a TV to your Homekit. This enables the possibility to change the input and the powerstate in one Homekit block. It also makes it possible to use the remote widget (or how it's called) to control the receiver and change the volume. This tv service is added to the plugin, so you can still use the switches if you want. An important note is that if you use the tv service, you have to add the them manually to your home app, after registering homebridge. This way it is possible to have more than one accessory which has remote widget support.

You can add multiple receivers by adding more than one `Device` in your config file. Every receiver must be registered manually ones after registering Homebridge. You can choose your input settings yourself. I added a list with all possible input settings I know of in Possible_inputs.json. You can add them all, but I only added the once I use, so the list stays short. It is also possible to change the name of the input to one with a personal touch ;).

A TV accessory has support for an info and a settings button. Normaly the settings button is accessible in the home app and the info button is accessible through the remote widget. As I don't use the info button but do use the settings button, I made it possible to switch the functionality of these buttons. This way you can have the settings menu under the `I` button in the remote widget. The volume control works in steps of 0.5. In the future, this must be update so you can hold the button to increase the volume by more.


### Switches
The standard switches, which are available for all different input types, are 'on' if the receiver is turned on and when the input is set to the specific switch. This allows you to create one switch for each input source and switch between input sources when the AVR is powered by just tapping the switch for that source. There is a polling loop embedded which checks every couple of seconds what input source is currently selected and will update the switches accordingly. For a short period of time it will appear like both inputs are "on". 

I add the option to make a general switch that polls for the state no matter the selected input. With this option, it is possible to have one main switch to turn on and off the receiver. You can still choose the default input when turning on the receiver with this switch. This functionality is configurable in the config with: `pollAllInput`.

### Demo images
<img src=https://raw.githubusercontent.com/Martvvliet/homebridge-denon-heos/master/images/SampleVid1.gif> <img src=https://raw.githubusercontent.com/Martvvliet/homebridge-denon-heos/master/images/SampleVid2.gif>

Left: Two times the same receiver as tv. Two separate switches added which also control a predefined input.
Right: When controlling the inputs with a tv tile, the separate switches are updated.

<img src=https://raw.githubusercontent.com/Martvvliet/homebridge-denon-heos/master/images/Sample_Inputs.png> <img src=https://raw.githubusercontent.com/Martvvliet/homebridge-denon-heos/master/images/Sample_Remote.png>

Left: The receiver added as tv. This enables the possibility to control the input and the power state in one tv block.
Right: When the receiver is added as tv, the receiver can be controlled with the remote. The volume buttons can be used to control the volumes. The info butten can be configured as info menu and settings menu.

<img src=https://raw.githubusercontent.com/Martvvliet/homebridge-denon-heos/master/images/Sample_Switches.png> <img src=https://raw.githubusercontent.com/Martvvliet/homebridge-denon-heos/master/images/Sample_Switches_2.png>

Left: A sample dedicated switch is added. This switch can be used to turn on the receiver and set it to the correct input.
Right: When settings the second dedicated switch, the input switches to Apple TV. The other switches is updated, and thus, turned off.




## Config

See sample-config.json for a complete sample json file. It is possible to add switches and tv services at the same time in one platform. The `pollInterval` is an optional value. Default is 3 seconds. If you want a lower or higher polling interval, set this value to a time in seconds. The following examples are given:

### TV Accessories
TV accessories are added as devices. The `switchInfoMenu` can be set to true if you want to switch the settings and info button functionality. Default is false. The inputs are automatically ordered alphabetically in homekit, so the order in the json doesn't matter. Check the `InputsSample.json` for the correct inputs ID's. `port` is optional and it's standard value is `"auto"`. If the plugin is not working, you can try to set it to `8080` for newer receivers, `80` for older ones and `"telnet"`for brand spanking new ones (first try auto as this automatically chooses the right ones). The found port, used when on auto, is visible in the homebridge log as: `port`.

```json
{
    "platforms": [{
        "platform": "DenonAVR",
        "pollInterval": 3,
        "devices": [{
            "name": "Denon Receiver",
            "ip": "192.168.1.45",
            "switchInfoMenu": true,
            "port": 8080,
            "inputs": [{
                "inputID": "MPLAY",
                "name": "Apple TV"
            },
            {
                "inputID": "GAME",
                "name": "iMac"
            },
            {
                "inputID": "TV",
                "name": "TV"
            },
            {
                "inputID": "AUX1",
                "name": "AUX"
            }]
        }]
    }]
}
```

### Switches
Set `pollAllInput` to true if you want a main switch to turn of the receiver no matter the selected input. Default is false. `port` is optional and it's standard value is `"auto"`. If the plugin is not working, you can try to set it to `8080` for newer receivers, `80` for older ones and `"telnet"`for brand spanking new ones (first try auto as this automatically chooses the right ones). The found port, used when on auto, is visible in the homebridge log as: `port`.
```json
{
    "platforms": [{
        "platform": "DenonAVR",
        "pollInterval": 3,
        "switches": [{
            "name": "AVR on Apple TV",
            "ip": "192.168.1.45",
            "port": 8080,
            "inputID": "MPLAY",
            "pollAllInput": false
        },
        {
            "name": "AVR on iMac",
            "ip": "192.168.1.45",
            "inputID": "GAME"
        }]
    }]
}
```

### Volume control
With volume control, you can set the volume level of your receiver. It adds a lightbulb which stands for the absolute volume level of your receiver. The volume limit is used as security. If you accidentally say to Siri: `Set receiver volume to 100`, your eardrums will at least survive. f the plugin is not working, you can try to set it to `8080` for newer receivers, `80` for older ones and `"telnet"`for brand spanking new ones (first try auto as this automatically chooses the right ones). The found port, used when on auto, is visible in the homebridge log as: `port`.
```json
{
    "platforms": [{
        "platform": "DenonAVR",
        "pollInterval": 3,
        "volumeControl": [{
            "name": "Receiver Volume",
            "ip": "192.168.1.45",
            "volumeLimit": 60,
            "port": 8080
        }]
    }]
}
```

### Choosing the correct inputIDs
If you are not sure what the name of the inputID is that you are using, you can use the following links to check what the current source inputID is. You need to change the ip address to your receivers IP and than paste it in your browser. `http://192.168.1.45:8080/goform/formMainZone_MainZoneXmlStatusLite.xml` or `http://192.168.1.45:80/goform/formMainZone_MainZoneXmlStatusLite.xml` in case of older AVR's. The result will look like the following:
```xml
<item>
  <Power>
    <value>ON</value>
  </Power>
  <InputFuncSelect>
    <value>SAT/CBL</value>
  </InputFuncSelect>
  <VolumeDisplay>
    <value>Absolute</value>
  </VolumeDisplay>
  <MasterVolume>
    <value>-50.0</value>
  </MasterVolume>
  <Mute>
    <value>off</value>
  </Mute>
</item>
```
You can see that it is currently on `SAT/CBL`. This means that you need as input (`"name"` is something you can change to something personal): 

```json
{
    "inputID": "SAT/CBL",
    "name": "Satellite/Cable"
}
```

## Possible errors
The following errors can be expected:
* `Some manual port number are not equal in config file with receiver: xx.xx.xx.xx.`
If you have multiple switches and/or tv accessories with the same IP but with different manual port settings, you get this error. Only set one port on one receiver.

* `Current port xx with ip: xx.xx.xx.xx, is not suitable. Use 80 or 8080 manually instead.`
The non-Heos receivers use `80` as port and the Heos receivers use `8080` as port. If you set a different port, you get this error.

* `Can't connect to receiver with ip: xx.xx.xx.xx and port: xx`
If this happens sometimes, it is not an issue. If this happens more often and you notice unresponsiveness in your home app, it is an issue. This might be due to a wrong IP or a wrong manual port number. It is also possible that the receiver is not accessible in the network. You can try to manuall acces your receiver with the following command in your browser: `http://xx.xx.xx.xx:xx/goform/formMainZone_MainZoneXmlStatusLite.xml`. The xx.xx.xx.xx is your IP and the xx is the manual port. If your network connection is the issue, try setting a static ip for your receiver an if you use wifi, try a lan cable.

* `Can not access receiver with IP: xx.xx.xx.xx. Might be due to a wrong port. Try 80 or 8080 manually in config file.`
It is possible that the the software was not able to extract the correct port from the available receivers. Make sure to connect the receivers to the network when (re)starting homebridge. If this doesn't fix the problem, try to set a manual port. `80` is for non-Heos models and `8080` is for Heos models.

* `Receiver with ip: xx.xx.xx.xx is disabled. Can't connect through http or Telnet.`
The specified receiver is not found in the network with the auto-discover function. This means that the auto port set will not work and the device information can not be set in homebridge. If you can control your device, you can ignore this warning. If the plugin is not working. You can try to set a manual port or set it to telnet and check if the plugin is working in the home app. Otherwise try checking the network connection of the receiver.

* `Can't connect to receiver with IP: xx.xx.xx.xx` 
The plugin can't connect to the receiver over Telnet. This might be due to a faulty network connection or something in the router. You can test this to open a manual Telnet connection with the receiver on port 23. Please Google if you don't know how this is possible.

* `Can't login at receiver with IP: xx.xx.xx.xx` 
The plugin can't connect to the receiver over Telnet. This might be due to a faulty network connection or something in the router. You can test this to open a manual Telnet connection with the receiver on port 23. Please Google if you don't know how this is possible.



## Further Reading and Thanks

Thanks to [nneubauers](https://github.com/nneubauer) for making a stable version that worked with newer Denon models like my AVR x1400. Thanks to [stfnhmplr](https://github.com/stfnhmplr), as I looked a bit to his Telnet implementation of the Denon. Also thanks to Jer G who took the time to inform me on the volume control of Denon receivers and input settings with special characters.

## Future work

* Improve volume control for remote widget.
* Add support for multi zone.
* Add default volume levels for specific inputs.
* Add option for auto discovery IP address.
* ~~Improve updating state of the multiple switches when changing one.~~
* ~~Improve polling code for more efficiency.~~
* ~~Add volume control with a Light Bulb for Siri volume control.~~
* ~~Add support for Telnet communication.~~
* ~~Auto discovery of communication method.~~


Changelog
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [Released]
## [2.3.1] - 2019-02-16
### Fixed
- Fix for older Node versions

## [2.3.0] - 2019-02-16
### Added
- Support for Telnet communication in case of newer Denon receivers. 

### Changed
- Changed port and protocol detection for better auto discovery functionality.

## [2.2.1] - 2019-02-04
### Added
- Option to use debug log through config

## [2.2.0] - 2019-02-03
### Added
- Added the possibility to add a lightbulb which stands for the volume level of the receiver.

### Fixed
- Fix with cached accessories of changed config.json

## [2.1.1] - 2019-02-01
### Fixed
- Fix for appearing devices in Home app.

## [2.1.0] - 2019-01-30
### Added
- When changing the state of a switch, the new state is pushed to your other switches and tv accessories of the same receiver. This means a faster state update if you switch input with switches. 

### Changed
- A more sophisticated version of the polling method is implemented. Now only one polling loop is running for every receiver you configured. 

### Fixed
- Fixes of previous versions.

## [2.0.8] - 2019-01-27
### Fixed
- Fix for DDos the receiver if you have several switches of the same receiver. 
- Overall crash fixes.

## [2.0.7] - 2019-01-26
### Fixed
- Fix in auto-discovery of port settings. 
- Improved stability manual setting of port. 
- Fixed bug that the settings of the legacy switches were resetted after resetting homebridge.

## [2.0.6] - 2019-01-19
### Fixed
- Fix for wrong use of inputName

## [2.0.5] - 2019-01-19
### Changed
- Removed some error messages.

## [2.0.4] - 2019-01-19
### Fixed
- Fix for auto detect port if not supported by receiver.

## [2.0.3] - 2019-01-19
### Fixed
- Fix for not recognising input if input has special characters.

## [2.0.2] - 2019-01-19
### Fixed
- Fix for wrongfully selecting upper input when changing Volume or using automations. 

## [2.0.1] - 2019-01-19
### Added
- Correct API port for older receivers. 
- Auto detect and manual support. 
- Extract info from AVR and add it to the homebridge device information. 

## [2.0.0] - 2019-01-10
### Added
- Initial release with support for TV accessories and multiple switches in one platform.

### Changed

### Removed

## [Unreleased]: 

Research/HTMLCommands.js
/* Menu items */
/* “Cursor Up” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNCUP
/* “Cursor Down" Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNCDN
/* “Cursor Left" Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNCLT
/* “Cursor Right” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNCRT
/* “Enter” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNENT
/* “RETURN” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNRTN
/* “GUI Menu ON” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNMEN ON
/* “UI Menu OFF” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNMEN OFF
/* “UI info” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MNINF


/* Volume items */
/* Mute */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MUON
/* Unmute */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MUOFF
/* “Volume Up” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MVUP
/* “Volume Down” Control */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MVDOWN
/* Set absolute volume level */
// http://192.168.1.45:8080/goform/formiPhoneAppDirect.xml?MV495


// http://192.168.1.45:8080/goform/formMainZone_MainZoneXmlStatusLite.xml

lib/discover.js
const request = require('request');
const parseString = require('xml2js').parseString;
const ssdp = require('node-ssdp').Client
, client = new ssdp({})

const mainFile = require('../index');

class discover {
	constructor(that, log, foundReceivers, autoDiscoverTime, pluginVersion) {
		this.foundReceivers = foundReceivers;

		client.on('notify', function () {
			log.error('Got a notification.')
		})
		
		var thisDiv = this;
		client.on('response', function inResponse(headers, code, rinfo) {
			let stringData = JSON.stringify(headers, null, '  ');
			if (stringData.match(/(d|D)enon/)) {
				// mainFile.logDebug('DEBUG: ' + stringData);
				let parseData = stringData.match(/"LOCATION": (.*?)(\d+\.\d+\.\d+\.\d+)(.*)/);
				let temp = parseData[0].split(/"/);

				let ipAddr = temp[3].replace("http://","");
				let colon = ipAddr.indexOf(":");
				ipAddr = ipAddr.substring(0,colon); 

				let receiver = [];

				receiver.push(ipAddr);
				receiver.push('0');
				receiver.push(temp[3]);
				receiver.push('Denon');
				receiver.push('homebridge-denon-heos');
				receiver.push('MVV123');
				receiver.push(pluginVersion);
				receiver.push(false);
				if (thisDiv.getIndexIP(ipAddr) == -1)
				{
					thisDiv.foundReceivers.push(receiver);
					thisDiv.retrieveDenonInformation(ipAddr, log);
					mainFile.logDebug('DEBUG: ' + ipAddr + ': ' + temp[3]);
				}
			}
		})
		client.search('ssdp:all')

		/* Stop search client */
		setTimeout(function () {
			client.stop()
			for (let i in that.configReceivers) {
				let tempAcces = that.configReceivers[i]
				if (!tempAcces.getDevInfoSet()) {
					if (tempAcces.hasManualPort()) {
						mainFile.logDebug('DEBUG: No Denon receiver with IP: ' + tempAcces.returnIP(), + ' found in network through auto-discovery. Using manual port: ' + tempAcces.returnPort());
					} 
					// else {
					// 	tempAcces.setDisableReceiver(true);
					// 	log.error('ERROR: No Denon receiver with IP: ' + tempAcces.returnIP() + ' found in network through auto-discovery. Check Denon network status or try setting a manual port.');
					// }
					tempAcces.setDevInfoSet(true);
				}
			}
		}, autoDiscoverTime)
	}

	getIndexIP(ipAddr) {
		for (var i in this.foundReceivers) {
			if (this.foundReceivers[i].indexOf(ipAddr) != -1)
				return i;
		}
		return -1;
	}

	getInfoAddress(ipAddr) {
		for (var i in this.foundReceivers) {
			if (this.foundReceivers[i][0].indexOf(ipAddr) != -1)
				return this.foundReceivers[i][2];
		}
		return null;
	}

	retrieveDenonInformation(ipAddr, log) {
		var index = this.getIndexIP(ipAddr);

		var that = this;
		request(this.getInfoAddress(ipAddr), function(error, response, body) {
			if(error) {
				mainFile.logDebug("Error while getting information of receiver with IP: " + that.foundReceivers[index][0]);
				mainFile.logDebug('DEBUG: ' + error);
			} else {
				body = body.replace(/:/g, '');
				parseString(body, function (err, result) {
					if(err) {
						mainFile.logDebug("Error while parsing retrieveDenonInformation. " + err);
					} else {
						try {
							try {
								that.foundReceivers[index][3] = result.root.device[0].manufacturer[0];
							} catch (error) {
								mainFile.logDebug('DEBUG: Fault in manufacturer: ' + result.root.device[0].manufacturer[0]);
							}
							try {
								that.foundReceivers[index][4] = (' ' + result.root.device[0].modelName[0]).slice(1);
							} catch (error) {
								mainFile.logDebug('DEBUG: Fault in modelName: ' + result.root.device[0].modelName[0]);
							}
							try {
								that.foundReceivers[index][5] = result.root.device[0].serialNumber[0];
							} catch (error) {
								mainFile.logDebug('DEBUG: Fault in serialNumber: ' + result.root.device[0].serialNumber[0]);
							}
							try {
								that.foundReceivers[index][1] = result.root.device[0].DMHX_WebAPIPort[0];
							} catch (error) {
								mainFile.logDebug('DEBUG: No WebAPIPort found, use Telnet.');
							}

							try {
								for (let i = 0; i < result.root.device[0].deviceList[0].device.length; i++){
									try {
										that.foundReceivers[index][6] = result.root.device[0].deviceList[0].device[i].firmware_version[0];
										break;
									} catch (error) {
									}
								}
							} catch (error) {
								mainFile.logDebug('DEBUG: No Firmware version found.');
							}

							that.foundReceivers[index][7] = true; // set information retrieved to true

							log('------------------------------');
							log('Receiver discovered in network:');
							log('IP Address: %s', that.foundReceivers[index][0]);
							if (that.foundReceivers[index][1] != '0')
								log('Port: %s', that.foundReceivers[index][1]);
							log('Model: %s', that.foundReceivers[index][4]);
							log('Serialnumber: %s', that.foundReceivers[index][5]);
							log('Firmware: %s', that.foundReceivers[index][6]);
							log('------------------------------');
						} catch (error) {
							mainFile.logDebug('DEBUG: Receiver with IP ' + ipAddr + ' not ready.');
							mainFile.logDebug('DEBUG: ' + error);
						}
					}
				});
			}
		});
	}

	setDenonInformation(that, log) {
		mainFile.logDebug('DEBUG: setDenonInformation: ' + that.ip + '. Device info set: ' + that.devInfoSet);

		let index;
		if (that.devInfoSet) {
			return true;
		} else {
			index = this.getIndexIP(that.ip);
			if (index == -1)
				return false
			if (!this.foundReceivers[index][7])
				return false
		}
		
		that.manufacturer = this.foundReceivers[index][3];
		that.modelName = this.foundReceivers[index][4];
		that.serialNumber = this.foundReceivers[index][5];
		that.firmwareRevision = this.foundReceivers[index][6];
		
		that.devInfoSet = true;
		// try {
		// 	if(this.foundReceivers[index][1].includes('80') && !that.usesManualPort)
		// 		that.webAPIPort = this.foundReceivers[index][1];
		// } catch (error) {
		// 	mainFile.logDebug('DEBUG: Error in discovered port number Error: ' + error);
		// }
		
		// try {
		// 	if(that.webAPIPort.includes('80'))
		// 		that.devInfoSet = true;
		// 	else {
		// 		log.error('ERROR: Current port ' + this.webAPIPort + ' with ip: ' + this.ip + ', is not suitable. Use 80 or 8080 manually instead.');
		// 		process.exit(22);
		// 	}
		// } catch (error) {
		// 	mainFile.logDebug('DEBUG: Error in webAPIPort. Not a correct port. Error: ' + error);
		// }
		return true;
	}
}

module.exports = discover;